### PR TITLE
T6927: adds option to set container name server

### PR DESCRIPTION
**Pull request by nvollmar**
## Change Summary
Add option to set name server of a container

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
https://vyos.dev/T6927

## Component(s) name
container

## Proposed changes
New config option to configure dns server.

## How to test
Set name server of a container
```
set container name <name> name-server <address>
```
Then verify `/etc/resolv.conf` within the container or `sudo podman inspect <name>`

## Smoketest result
```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_container.py
test_api_socket (__main__.TestContainer.test_api_socket) ... ok
test_basic (__main__.TestContainer.test_basic) ... ok
test_cpu_limit (__main__.TestContainer.test_cpu_limit) ... ok
test_dual_stack_network (__main__.TestContainer.test_dual_stack_network) ... ok
test_ipv4_network (__main__.TestContainer.test_ipv4_network) ... ok
test_ipv6_network (__main__.TestContainer.test_ipv6_network) ... ok
test_name_server (__main__.TestContainer.test_name_server) ... ok
test_network_mtu (__main__.TestContainer.test_network_mtu) ... ok
test_no_name_server (__main__.TestContainer.test_no_name_server) ... ok
test_uid_gid (__main__.TestContainer.test_uid_gid) ... ok
```

## Checklist:
- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [X] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Review by sever-sever — APPROVED** [review]

**Comment by sever-sever** [comment]
Smoketest fails
```
DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_container.py
DEBUG - test_api_socket (__main__.TestContainer.test_api_socket) ... ok
DEBUG - test_basic (__main__.TestContainer.test_basic) ... ok
DEBUG - test_cpu_limit (__main__.TestContainer.test_cpu_limit) ... ok
DEBUG - test_dual_stack_network (__main__.TestContainer.test_dual_stack_network) ... ok
DEBUG - test_ipv4_network (__main__.TestContainer.test_ipv4_network) ... ok
DEBUG - test_ipv6_network (__main__.TestContainer.test_ipv6_network) ... ok
DEBUG - test_name_server (__main__.TestContainer.test_name_server) ... ERROR
DEBUG - test_name_server (__main__.TestContainer.test_name_server) ... FAIL
DEBUG - test_network_mtu (__main__.TestContainer.test_network_mtu) ... FAIL
DEBUG - test_no_name_server (__main__.TestContainer.test_no_name_server) ... FAIL
DEBUG - test_uid_gid (__main__.TestContainer.test_uid_gid) ... FAIL
DEBUG - 
DEBUG - ======================================================================
DEBUG - ERROR: test_name_server (__main__.TestContainer.test_name_server)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_container.py", line 106, in test_name_server
DEBUG -     self.cli_commit()
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/base_vyostest_shim.py", line 86, in cli_commit
DEBUG -     self._session.commit()
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 264, in commit
DEBUG -     out = self.__run_command([COMMIT])
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 201, in __run_command
DEBUG -     raise ConfigSessionError(output)
DEBUG - vyos.configsession.ConfigSessionError: [ container ]
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/services/vyos-configd", line 137, in run_script
DEBUG -     script.apply(c)
DEBUG -   File "/usr/libexec/vyos//conf_mode/container.py", line 505, in apply
DEBUG -     cmd(f'systemctl restart vyos-container-{name}.service')
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/utils/process.py", line 155, in cmd
DEBUG -     raise OSError(code, feedback)
DEBUG - PermissionError: [Errno 1] failed to run command: systemctl restart vyos-container-dns-test.service
DEBUG - returned:
DEBUG - exit code: 1
DEBUG - 
DEBUG - [[container]] failed
DEBUG - Commit failed
DEBUG - 
DEBUG - 
DEBUG - ======================================================================
DEBUG - FAIL: test_name_server (__main__.TestContainer.test_name_server)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_container.py", line 71, in tearDown
DEBUG -     self.assertEqual(units, [])
DEBUG - AssertionError: Lists differ: ['/run/systemd/system/vyos-container-dns-test.service'] != []
DEBUG - 
DEBUG - First list contains 1 additional elements.
DEBUG - First extra element 0:
DEBUG - '/run/systemd/system/vyos-container-dns-test.service'
DEBUG - 
DEBUG - - ['/run/systemd/system/vyos-container-dns-test.service']
DEBUG - + []
DEBUG - 
DEBUG - ======================================================================
DEBUG - FAIL: test_network_mtu (__main__.TestContainer.test_network_mtu)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_container.py", line 71, in tearDown
DEBUG -     self.assertEqual(units, [])
DEBUG - AssertionError: Lists differ: ['/run/systemd/system/vyos-container-dns-test.service'] != []
DEBUG - 
DEBUG - First list contains 1 additional elements.
DEBUG - First extra element 0:
DEBUG - '/run/systemd/system/vyos-container-dns-test.service'
DEBUG - 
DEBUG - - ['/run/systemd/system/vyos-container-dns-test.service']
DEBUG - + []
DEBUG - 
DEBUG - ======================================================================
DEBUG - FAIL: test_no_name_server (__main__.TestContainer.test_no_name_server)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_container.py", line 71, in tearDown
DEBUG -     self.assertEqual(units, [])
DEBUG - AssertionError: Lists differ: ['/run/systemd/system/vyos-container-dns-test.service'] != []
DEBUG - 
DEBUG - First list contains 1 additional elements.
DEBUG - First extra element 0:
DEBUG - '/run/systemd/system/vyos-container-dns-test.service'
DEBUG - 
DEBUG - - ['/run/systemd/system/vyos-container-dns-test.service']
DEBUG - + []
DEBUG - 
DEBUG - ======================================================================
DEBUG - FAIL: test_uid_gid (__main__.TestContainer.test_uid_gid)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_container.py", line 71, in tearDown
DEBUG -     self.assertEqual(units, [])
DEBUG - AssertionError: Lists differ: ['/run/systemd/system/vyos-container-dns-test.service'] != []
DEBUG - 
DEBUG - First list contains 1 additional elements.
DEBUG - First extra element 0:
DEBUG - '/run/systemd/system/vyos-container-dns-test.service'
DEBUG - 
DEBUG - - ['/run/systemd/system/vyos-container-dns-test.service']
DEBUG - + []
DEBUG - 
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 10 tests in 149.984s
```

**Comment by nvollmar** [comment]
Yes, as noted in the PR I'm currently having issues building an image to debug/test. So I created this PR as draft until I can get things final

File: src/conf_mode/container.py [hunk 1]
```diff
@@ -363,9 +363,14 @@ def generate_run_arguments(name, container_config):
     if 'allow_host_pid' in container_config:
       host_pid = '--pid host'
 
+    name_server = ''
+    if 'name_server' in container_config:
+        for ns in container_config['name_server']:
+            name_server += f'--dns {ns}'
```
Comment (sever-sever) [inline src/conf_mode/container.py:369]: Can we add 2 name-servers?

Comment (nvollmar) [reply src/conf_mode/container.py:369]: yes, you can specify multiple name servers. I've tested with 2, both were added to the container.

**Review by sever-sever — APPROVED** [review]

File: smoketest/scripts/cli/test_container.py [hunk 2]
```diff
@@ -96,6 +96,18 @@ def test_basic(self):
...
+        name_server = '192.168.0.1'
+        self.cli_set(base_path + ['name', cont_name, 'allow-host-networks'])
+        self.cli_set(base_path + ['name', cont_name, 'image', cont_image])
+        self.cli_set(base_path + ['name', cont_name, 'name-server', name_server])
+
+        self.cli_commit()
+
+        n = cmd_to_json(f'sudo podman inspect {cont_name}')
+        self.assertEqual(n['HostConfig']['Dns'][0], name_server)
+
```
Comment (c-po) [inline smoketest/scripts/cli/test_container.py:110]: Just beeing super correct here, can you please also read /etc/resolv.conf from within the container to be 100% sure?

This file should have an example hos to acces fles insode a container.

Other then that tiny thing, LGTM

**Review by c-po — CHANGES_REQUESTED** [review]

File: smoketest/scripts/cli/test_container.py (hunk 2 above)
Comment (nvollmar) [reply smoketest/scripts/cli/test_container.py:110]: added

**Review by c-po — APPROVED** [review]

**Comment by nvollmar** [comment]
I need to add a validation, just discovered `--dns` won't have an effect if it is attached to a container network with dns enabled. If container network is used it needs to have `--disable-dns` for it to work.

**Pull request merged**